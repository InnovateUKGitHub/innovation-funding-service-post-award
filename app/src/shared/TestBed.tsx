import { createMemoryHistory } from "history";
import _merge from "lodash.merge";
import { Router } from "react-router-dom";
import { Copy } from "@copy/Copy";
import { mountedContext } from "@ui/components/atomicDesign/atoms/providers/Mounted/Mounted";
import { PageTitleProvider } from "@ui/features/page-title";
import { ContentProvider } from "@ui/redux/contentProvider";
import { IStores, StoresProvider } from "@ui/redux/storesProvider";
import { ClientConfigProvider } from "@ui/components/providers/ClientConfigProvider";
import { IClientConfig } from "src/types/IClientConfig";
import { AccEnvironment } from "@framework/constants/enums";

export type TestBedStore = Partial<IStores>;

export interface ITestBedProps {
  children: React.ReactElement;
  competitionType?: string;
  stores?: TestBedStore;
  isServer?: boolean;
  pageTitle?: string;
  /**
   * `shouldOmitRouterProvider` will prevent a router provider being added to the test bed.
   * Because there can only be one router provider in the tree, it may happen that in
   * some cases there will need to be a separate router provider, in which case pass this flag
   * in to enable building.
   */
  shouldOmitRouterProvider?: boolean;
  extendClientConfig?: (config: RecursiveMutable<IClientConfig>) => void;
}

/**
 * TestBed is a component configured with all stores needed for tests to run
 *
 * Note: When testing a component that consumes <Content />, it expects a provider to be present.
 */
export function TestBed({
  isServer = false,
  stores,
  competitionType,
  children,
  pageTitle = "stub-displayTitle",
  shouldOmitRouterProvider,
  extendClientConfig,
}: ITestBedProps) {
  const stubStores = {
    users: {
      getCurrentUser: () => ({ csrf: "stub-csrf" }),
    },
    navigation: {
      getPageTitle: () => ({
        displayTitle: "stub-displayTitle",
      }),
    },
  };

  const storesValue = _merge(stubStores, stores) as Required<TestBedStore>;

  const clientConfig = {
    features: {
      changePeriodLengthWorkflow: false,
      customContent: false,
      searchDocsMinThreshold: 9999,
      futureTimeExtensionInYears: 5,
    },
    options: {
      maxFileSize: 33554432,
      maxUploadFileCount: 10,
      maxFileBasenameLength: 80,
      permittedFileTypes: [
        "pdf",
        "xps",
        "doc",
        "docx",
        "rtf",
        "txt",
        "csv",
        "odt",
        "ppt",
        "pptx",
        "odp",
        "xls",
        "xlsx",
        "ods",
        "jpg",
        "jpeg",
        "png",
        "odg",
      ],
      permittedTypes: {
        pdfTypes: ["pdf", "xps"],
        textTypes: ["doc", "docx", "rtf", "txt", "csv", "odt"],
        presentationTypes: ["ppt", "pptx", "odp"],
        spreadsheetTypes: ["xls", "xlsx", "ods"],
        imageTypes: ["jpg", "jpeg", "png", "odg"],
      },
      bankCheckValidationRetries: 3,
      bankCheckAddressScorePass: 6,
      bankCheckCompanyNameScorePass: 6,
      standardOverheadRate: 20,
      numberOfProjectsToSearch: 3,
      maxClaimLineItems: 120,
      nonJsMaxClaimLineItems: 10,
    },
    ifsRoot: "https://ifs-accdev.apps.ocp4.org.innovateuk.ukri.org",
    ssoEnabled: false,
    logLevel: "VERBOSE",
    accEnvironment: AccEnvironment.BASE,
  };

  // Modify the clientConfig if the TextBed wants to.
  extendClientConfig?.(clientConfig);

  // Note: We need a way of upfront toggling this can use 'MountedProvider'
  const testBedMountState = { isServer, isClient: !isServer };

  const history = createMemoryHistory();

  const Providers = (
    <mountedContext.Provider value={testBedMountState}>
      <ClientConfigProvider config={clientConfig}>
        <PageTitleProvider title={pageTitle}>
          <StoresProvider value={storesValue}>
            <ContentProvider value={new Copy({ competitionType })}>{children}</ContentProvider>
          </StoresProvider>
        </PageTitleProvider>
      </ClientConfigProvider>
    </mountedContext.Provider>
  );

  return shouldOmitRouterProvider ? (
    Providers
  ) : (
    <Router location={history.location} navigator={history}>
      {Providers}
    </Router>
  );
}

export type HookTestBedProps = Omit<ITestBedProps, "children">;

export const hookTestBed = (props: HookTestBedProps) => ({
  wrapper: (wrapperProps: ITestBedProps) => <TestBed {...props} {...wrapperProps} />,
});

export default TestBed;
