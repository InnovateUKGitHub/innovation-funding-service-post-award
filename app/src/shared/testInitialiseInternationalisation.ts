import { CopyLanguages, CopyNamespaces } from "@copy/data";
import i18next, { ResourceKey } from "i18next";

const testInitialiseInternationalisation = <T extends ResourceKey>(
  copy: T,
  additionalCopy?: { [key: string]: ResourceKey },
) => {
  return i18next.init({
    lng: CopyLanguages.en_GB,
    ns: CopyNamespaces.DEFAULT,
    fallbackLng: CopyLanguages.en_GB,
    fallbackNS: CopyNamespaces.DEFAULT,
    resources: {
      [CopyLanguages.en_GB]: {
        [CopyNamespaces.DEFAULT]: copy,
        ...additionalCopy,
      },
    },
  });
};

export { testInitialiseInternationalisation };
