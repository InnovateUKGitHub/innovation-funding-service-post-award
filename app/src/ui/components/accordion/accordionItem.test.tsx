import userEvent from "@testing-library/user-event";
import { render, act } from "@testing-library/react";

import { TestBed } from "@shared/TestBed";
import { AccordionItem, AccordionItemProps } from "..";
import { testInitialiseInternationalisation } from "@shared/testInitialiseInternationalisation";

const stubContent = {
  example: {
    contentTitle: "stub-exampleContentTitle",
  },
};

const defaultProps: AccordionItemProps = {
  title: "stub-title-as-string",
  qa: "stub-qa",
  children: <p>stub-children</p>,
};

const setup = (props?: Partial<AccordionItemProps>, isServer?: boolean) => {
  const rtl = render(
    <TestBed isServer={isServer}>
      <AccordionItem {...defaultProps} {...props} />
    </TestBed>,
  );

  const getAccordion = (textToQuery: string) => {
    const titleNode = rtl.getByText(textToQuery);
    const toggleContainer = titleNode.parentElement;

    if (!toggleContainer) throw Error(`Can't find container for '${textToQuery}'`);

    const accordionNode = toggleContainer.parentElement;

    if (!accordionNode) throw Error(`Can't find accordion node for '${textToQuery}'`);

    return { titleNode, accordionNode };
  };

  const queryOpenAccordions = () => {
    return rtl.container.querySelectorAll("govuk-accordion__section--expanded");
  };

  const queryAccordionToggle = () => {
    return rtl.queryByTestId("accordion-toggle");
  };

  return {
    ...rtl,
    getAccordion,
    queryOpenAccordions,
    queryAccordionToggle,
  };
};

describe("AccordionItem", () => {
  beforeAll(async () => {
    await testInitialiseInternationalisation(stubContent);
  });

  describe("@renders", () => {
    describe("with isOpen", () => {
      test.each`
        name            | isOpen
        ${"when true"}  | ${true}
        ${"when false"} | ${false}
      `("$name", ({ isOpen }) => {
        const { container } = setup({ isOpen });

        expect(container.firstChild).toMatchSnapshot();
      });
    });

    describe("with a title", () => {
      test("with title as string", () => {
        const stubTitle = "stub-title-as-string";
        const { queryByText } = setup({ title: stubTitle });

        expect(queryByText(stubTitle)).toBeInTheDocument();
      });

      test("with title as content solution", () => {
        const { getAccordion } = setup({ title: x => x.example.contentTitle });

        const { titleNode, accordionNode } = getAccordion(stubContent.example.contentTitle);

        expect(titleNode).toBeInTheDocument();
        expect(accordionNode.tagName).toBe("BUTTON");
      });
    });

    describe("when mounted", () => {
      test("when client should render a button", () => {
        const { queryAccordionToggle } = setup();

        expect(queryAccordionToggle()).toBeInTheDocument();
      });

      test("when client should render with a span button", () => {
        const { queryAccordionToggle } = setup(undefined, true);

        expect(queryAccordionToggle()).not.toBeInTheDocument();
      });
    });
  });

  describe("@calls", () => {
    test("with onClick when clicked", async () => {
      const stubToggleFn = jest.fn();
      const { queryAccordionToggle } = setup({ onClick: stubToggleFn });
      const toggleBtn = queryAccordionToggle();

      if (!toggleBtn) throw Error("Toggle button was not found!");

      expect(stubToggleFn).toHaveBeenCalledTimes(0);

      await act(() => userEvent.click(toggleBtn));

      expect(stubToggleFn).toHaveBeenCalledTimes(1);
    });
  });
});
