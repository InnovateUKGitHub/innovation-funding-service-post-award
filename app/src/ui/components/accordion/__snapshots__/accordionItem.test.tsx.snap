// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionItem @renders with isOpen when false 1`] = `
<div
  class="govuk-accordion__section"
  data-qa="stub-qa"
>
  <div
    class="govuk-accordion__section-header"
  >
    <h2
      class="govuk-heading-l govuk-accordion__section-heading"
    >
      <button
        aria-controls="accordion-default-content-stub-qa"
        aria-expanded="false"
        class="govuk-accordion__section-button"
        data-qa="accordion-toggle"
        type="button"
      >
        <span
          class="govuk-accordion__section-heading-text"
          id="accordion-default-heading-stub-qa"
        >
          <span
            class="govuk-accordion__section-heading-text-focus"
          >
            stub-title-as-string
          </span>
        </span>
        <span
          class="govuk-visually-hidden govuk-accordion__section-heading-divider"
        >
          , 
        </span>
        <span
          class="govuk-accordion__section-toggle"
        >
          <span
            class="govuk-accordion__section-toggle-focus"
          >
            <span
              class="govuk-accordion-nav__chevron govuk-accordion-nav__chevron--down"
            />
            <span
              class="govuk-accordion__section-toggle-text"
            >
              Show
               
              <span
                class="govuk-visually-hidden"
              >
                this section
              </span>
            </span>
          </span>
        </span>
      </button>
    </h2>
  </div>
  <div
    aria-labelledby="accordion-default-heading-stub-qa"
    class="govuk-accordion__section-content"
    id="accordion-default-content-stub-qa"
  >
    <p>
      stub-children
    </p>
  </div>
</div>
`;

exports[`AccordionItem @renders with isOpen when true 1`] = `
<div
  class="govuk-accordion__section govuk-accordion__section--expanded"
  data-qa="stub-qa"
>
  <div
    class="govuk-accordion__section-header"
  >
    <h2
      class="govuk-heading-l govuk-accordion__section-heading"
    >
      <button
        aria-controls="accordion-default-content-stub-qa"
        aria-expanded="true"
        class="govuk-accordion__section-button"
        data-qa="accordion-toggle"
        type="button"
      >
        <span
          class="govuk-accordion__section-heading-text"
          id="accordion-default-heading-stub-qa"
        >
          <span
            class="govuk-accordion__section-heading-text-focus"
          >
            stub-title-as-string
          </span>
        </span>
        <span
          class="govuk-visually-hidden govuk-accordion__section-heading-divider"
        >
          , 
        </span>
        <span
          class="govuk-accordion__section-toggle"
        >
          <span
            class="govuk-accordion__section-toggle-focus"
          >
            <span
              class="govuk-accordion-nav__chevron"
            />
            <span
              class="govuk-accordion__section-toggle-text"
            >
              Hide
               
              <span
                class="govuk-visually-hidden"
              >
                this section
              </span>
            </span>
          </span>
        </span>
      </button>
    </h2>
  </div>
  <div
    aria-labelledby="accordion-default-heading-stub-qa"
    class="govuk-accordion__section-content"
    id="accordion-default-content-stub-qa"
  >
    <p>
      stub-children
    </p>
  </div>
</div>
`;
